include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,237 @@
+			const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+			const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
 #endif
 
+			const float NdotH = microfacet.isotropic.NdotH;
+			const float NdotH2 = microfacet.isotropic.NdotH2;
+#ifndef ALL_ISOTROPIC_BXDFS
+			const float TdotH2 = microfacet.TdotH * microfacet.TdotH;
+			const float BdotH2 = microfacet.BdotH * microfacet.BdotH;
+#endif
+
+			BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+				CASE_BEGIN(ndf, NDF_GGX) {
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG_wo_clamps(NdotH2, NdotL, NdotL2, NdotV, NdotV2, a2);
 #else
-			bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_DG(currBSDFParams, currInteraction, a, a2, ay, ay2);
+					bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG_wo_clamps(NdotH2, TdotH2, BdotH2, NdotL, NdotL2, TdotL2, BdotL2, NdotV, NdotV2, TdotV2, BdotV2, ax, ax2, ay, ay2);
+#endif
+				} CASE_END
 #endif
 
-#ifndef ONLY_ONE_NDF
-		} else
+#ifdef NDF_BECKMANN
+				CASE_BEGIN(ndf, NDF_BECKMANN) {
+#ifdef ALL_ISOTROPIC_BXDFS
+					bxdf_eval_scalar_part = irr_glsl_beckmann_height_correlated_cos_eval_DG_wo_clamps(NdotH2, NdotL2, NdotV2, a2);
+#else
+					bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG_wo_clamps(NdotH2, TdotH2, BdotH2, NdotL2, TdotL2, BdotL2, NdotV2, TdotV2, BdotV2, ax, ax2, ay, ay2);
 #endif
+				} CASE_END
 #endif
 
 #ifdef NDF_PHONG
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_PHONG) {
-#endif
-
-
-			float n = irr_glsl_alpha2_to_phong_exp(a2);
+				CASE_BEGIN(ndf, NDF_PHONG) {
+					float n = irr_glsl_alpha2_to_phong_exp(a2);
 #ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, n, a2);
+					bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotV2, NdotL2, n, a2);
 #else
-			float ny = irr_glsl_alpha2_to_phong_exp(ay2);
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
-#endif
-
-
-#ifndef ONLY_ONE_NDF
-		} else
+					float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+					bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG_wo_clamps(NdotH, NdotH2, TdotH2, BdotH2, TdotL2, BdotL2, TdotV2, BdotV2, NdotV2, NdotL2, nx, ny, ax2, ay2);
 #endif
+				} CASE_END
 #endif
 
-#ifndef ONLY_ONE_NDF
-		{} //else "empty braces"
-#endif
-	}
+				CASE_OTHERWISE
+				{} //else "empty braces"
+			END_CASES
 
-	uvec3 regs = instr_decodeRegisters(instr);
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
-#endif
+				float VdotH = microfacet.isotropic.VdotH;
+				float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
```

Comment:
`VdotH` does not need to be clamped... `VdotH<FLT_MIN` is simply an invalid microfacet normal (was checked by `is_valid` before anyway OR guaranteed by constraints on NdotV and NdotL for BRDFs)

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/554#discussion_r527890123